code.js

```diff
@@ -1,3 +1,41 @@
 function mergesort(array) {
+    let currentSize;
+    let leftStart
+    let arraySize = array.length
+
+    for (let currentSize = 1; currentSize <= arraySize -1; currentSize = currentSize * 2) {
+        for (let leftStart = 0; leftStart < arraySize; leftStart += 2 * currentSize) {
+            let mid = Math.min(leftStart + currentSize - 1, arraySize - 1);
+            let rightEnd = Math.min(leftStart + 2 * currentSize - 1, arraySize - 1);
+            merge(array, leftStart, mid, rightEnd)
+        }
+    }
     return array;
 }
+function merge(array, left, mid, right){
+
+let x = mid + 1;
+if(array[mid] <= array[x]){
+    return;
+}
+    while (left <= mid && x <= right){
+        if(array[left] <= array[x]){
+            left++;
+        }
+        else{
+            let value = array[x];
+            let index = x;
+
+            while(index != left){
```

Comment:
This loop is nested inside another one -- does that affect complexity?

Reply:
Yeah, the presence of a nested loop will affect the time complexity of the algorithm but the time complexity isn't always the product of the individual loops. The outerloop runs $logn$ times because the size of the sub-arrays being merged double with each iteration. So, the number of iterations is proportional to the base_2 logarithm of that array size. The innerloop is actually just merging the sub-arrays and is processes once for each level

Reply:
So what's the complexity of the `merge()` function? And what is it for the not in place mergesort?

Reply:
The complexity of the merge function is $O(n)$ because the function contains a single while loop that processes each element once.
The not in place merge sort is also $O(nlogn)$ but it needs more space proportional the its array size. The in place version does not require more space then the input array

Reply:
The outer loop iterates over all elements. What does the inner loop do? Does that not add to the complexity?